Handle non-Linux systems, build docker outside of source tree, replace & delete per-component Makefiles and push scripts

mattdelco opened the pull request:
This PR is being used to achieve the following:

* Permit/handle builds [i.e., cross-compiles] for other OSes.  go/bin is mostly left unused besides dep, golint, and a few sentinel files. artifacts placed in go/out/[lx|mac|win]/$GOARCH/[debug|release] .  To help consumers of the build (e.g., tests) find the directory (and to help insulate them from future changes in the Makefile) I've toyed with the prospect of providing targets that output the directory location (E.g., script can use ISTIO_OUT=$(make where-is-out) ).
* Docker builds now performed under the go/out/... directory, rather than scribbling in the source tree.  Currently I've punted on docker builds for non-Linux OSes.  The targets are exposed only when compiling for Linux (e.g., "make docker" will result in "unknown target" on a Mac, though "make GOOS=linux docker" should try to build [whether it actually works is TBD]).  I've modified 
  security/bin/gen-keys.sh to permit output to another location.
* Teaches make how to create the installer tars.  Make can handle more of the end-to-end build.
  The current targets are "make docker.save sidecar.deb istio-archive", though a single target
  could be made an alias for all of them.  The release build handles push/upload in a separate step though an integrated push is more like "make HUB=gcr.io/foo TAG=0.5.0 push".
* check go version with fewer issues in half the time compared to the current script.
* Teaches make how to tag and/or push the docker images it builds.

The PR now works with the presumit tests and release builds but will likely break other uses (e.g., people who use certain scripts or per-component Makefiles, but they can migrate to the top-level Makefile [assuming I've properly captured the handling for that use case]).  Given the potential for disruption (and timing of the monthly release) I don't intent for this to get committed soon.

Deletes files that ideally won't be needed anymore (though no doubt there's some unhandled issues):
bin/docker_lib.sh
bin/push
bin/verify_go_version.sh
mixer/Makefile
mixer/bin/push-docker
mixer/example/servicegraph/Makefile
pilot/Makefile
pilot/bin/push-debian.sh
pilot/bin/push-docker
pilot/bin/upload-istioctl
security/Makefile
security/bin/check_build_files.sh
security/bin/push-debian.sh
security/bin/push-docker
misc ".gitignore" files that were added to hide the scribbling in the source tree.

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 4e799b9aeac1bbb04b9511d314da262574cbb5c8 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3275/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 5ba4ccdd07afe08ea0b35a1fd27c2e103747b042 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3276/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 6a5102233a55418914364949110487ba6bfee4f3 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3279/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | b601b9ae8d3cbad225b5054589378e1109acce42 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3280/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | e5fd3b13bfa36f30143b8f8860af3ddf0dc72b83 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3282/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | e08b04ce6aa4c0912b18affc39d40b7cb713ba9c | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3291/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 55a918fa8309f1d5659fd42523fed4b1c549105b | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3293/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | d95f7d0d1ef98361f85121ae3e0df208e076e4bf | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3295/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | f2a5c497008aa4664711a8882e8138b760c04171 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3296/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | c82f1f0237c20ac0e1539cc81b8c1ce0acd2fd8a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3313/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | cde1d788b14f086f5b55cb201644d97bb20e913a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/1801/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | cde1d788b14f086f5b55cb201644d97bb20e913a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/1801/) | `/test istio-pilot-e2e`
prow/e2e-smoke.sh | cde1d788b14f086f5b55cb201644d97bb20e913a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/1834/) | `/test e2e-smoke`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | cde1d788b14f086f5b55cb201644d97bb20e913a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/1801/) | `/test istio-pilot-e2e`
prow/e2e-smoke.sh | cde1d788b14f086f5b55cb201644d97bb20e913a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/1834/) | `/test e2e-smoke`
prow/istio-presubmit.sh | d3f642606b01813ee192918830ad2cf654cfd0c2 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3317/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-smoke.sh | cde1d788b14f086f5b55cb201644d97bb20e913a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/1834/) | `/test e2e-smoke`
prow/istio-pilot-e2e.sh | 1239cbb59e70246e9da0402d4b732266e07aa9a7 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/1802/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | 9e909ef9f00ae8c90eb39bc37624efdf514726f5 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/1846/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

rshriram commented:
While you are at this, would you be interested in doing this: https://gist.github.com/prwhite/8168133 ? Or atleast for the targets you are adding?

Also, thanks for the cleanups. There were a couple more things in my old PR #2674 (its incomplete anyway). If you could subsume those changes here, I can close that PR as well.

rshriram commented:
Second suggestion: Can you push this to a branch in istio/istio, so that we can try to compile on a Mac and make sure this works? if not, we could chime in and add fixes/point you to the errors at the very least.

rshriram commented on Makefile:
```diff
@@ -357,204 +449,55 @@ racetest: pilot-racetest mixer-racetest security-racetest broker-racetest galley
-# create a DOCKER_TAR_TARGETS that's each of DOCKER_TARGETS with a tar. prefix
-DOCKER_TAR_TARGETS:=
-$(foreach TGT,$(DOCKER_TARGETS),$(eval DOCKER_TAR_TARGETS+=tar.$(TGT)))
+include tools/istio-docker.mk
```
can we put these in bin/ ? Tools has some code as well. bin is all scripts.

mattdelco replied:
When Costin requested I put the docker stuff in a separate mk he suggested this name/location.  These mk aren't quite executable either so should we start a "make/"?  Other possible candidates to move to this directory (if used) would be tests/istio.mk and tools/deb/istio.mk.

rshriram commented on Makefile:
```diff
@@ -357,204 +449,55 @@ racetest: pilot-racetest mixer-racetest security-racetest broker-racetest galley
 
-push: checkvars clean.installgen installgen
-	$(ISTIO_GO)/bin/push $(HUB) $(TAG) $(GS_BUCKET)
+push.istioctl-all: istioctl-all
```
and this is better represented as gcs.push.istioctl-all ..

mattdelco replied:
Honestly I wouldn't mind deleting it.  The functionality seems a bit niche.  However, I'm just trying to maintain compatibility with what bin/push performs though I'm not certain of the need or use case.  Having said that, I will do the rename.

rshriram commented on tools/istio-docker.mk:
```diff
@@ -0,0 +1,209 @@
+docker: docker.all
+
+# Build docker images for pilot, mixer, ca using prebuilt binaries
+docker.prebuilt: docker.tag
```
You can remove this target and simply update the .circleci/config.yaml file to point to the new targets..

mattdelco replied:
I was updating .circleci/config.yml to reflect your other PR (change "make go-build" to "make build") but there seems to be something odd about the line endings in the file--emacs thinks it's seeing DOS [i.e., Windows] line endings.  I don't think I want be listed as modifying the whole file so I'll just stick to the lines I'm touching.

rshriram commented on prow/istio-pilot-e2e.sh:
```diff
@@ -56,4 +56,4 @@ HUB="gcr.io/istio-testing"
 cd ${GOPATH}/src/istio.io/istio
 ./bin/init.sh
-make -C "${ROOT}/pilot" e2etest HUB="${HUB}" TAG="${GIT_SHA}" TESTOPTS="-mixer=false -use-sidecar-injector=false -use-admission-webhook=false"
+make e2e_pilot HUB="${HUB}" TAG="${GIT_SHA}" TESTOPTS="-mixer=false -use-sidecar-injector=false -use-admission-webhook=false"
```
shouldn't this change be reflected in circleci as well?

mattdelco replied:
I'm not sure I understand the question, but I read it as suggesting that circleCI can be changed to use "make e2e_pilot" rather than launch bin/pilot-integration-test directly.  I can give that a try.  It looks like prow and circleCI use a different set of parameters, though I don't know if that's intentional or accidental (circleCI uses true for -mixer while prow uses -false.  CircleCI also has different sections for -auth with enable vs disable and prow doesn't specify anything).

rshriram replied:
Prow should use mixer true. And yes I meant we should sync them up.

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | c3126f73b98f6ced870b5dc8bfd2ca23bce53d89 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3456/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

mattdelco commented:
For the record, the unrelated experiment of using code owners is what caused the automatic addition of the long list of requested reviewers (i.e., I didn't do it explicitly).

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 8801a5dcbbe5ea56e6c5b61b41c37f0f09e8fab1 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3499/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | bda3601f325870e44a418293df3750bf68e996e6 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3502/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

rshriram commented:
That’s fine matt. And it’s intentional since everyone is affected by the code in tests directory and should pay attention. 

With regard to circleci, I too saw this CRLF thing. is it possible to edit the file without doing the dos2Unix thing?

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | b4b7b6b7338ff638d8cd92a3da4ac434f3b49290 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3507/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

mattdelco commented:
For the circleCI file I believe I've only modified the intended lines rather than dos2Unix the whole file, and I think me editor is respecting the oddness of the file and therefore the file continues to have consistent line endings.

As for Mac, I'm not sure I have enough git skill yet (nor appropriate repo permissions) to create a branch on the istio repo.  I did checkout the source on my Mac ("master6b" branch on mattdelco/istio, which I suppose anyone else can do too) and the guinea pig I picked ("make pilot") seemed to build as expected.

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | bf00be38583b3ee85b598f54c62d14d05d198666 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/1921/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

costinm commented:
Every PR is auto-creating a branch, it can be checked out (and modified) by
other users.

We should do a separate PR to revert the yaml file to \n, someone on
windows did an edit...

Re. istio.mk: I'm not sure what's right, on one side I don't think we
should end with a complex recursive make
with many files, on the other side some pieces are relatively standalone (
like creating the deb, and
running the e2e tests ). I think it's valuable for the main makefile to
have the 'core' targets and be somehow
simple to read - while 'hiding' trickier stuff in a helper mk file, and
tools seems a good place. But we can iterate
and find what works.

On Fri, Jan 26, 2018 at 8:33 PM, mattdelco <notifications@github.com> wrote:

> For the circleCI file I believe I've only modified the intended lines
> rather than dos2Unix the whole file, and I think me editor is respecting
> the oddness of the file and therefore the file continues to have consistent
> line endings.
>
> As for Mac, I'm not sure I have enough git skill yet (nor appropriate repo
> permissions) to create a branch on the istio repo. I did checkout the
> source on my Mac ("master6b" branch on mattdelco/istio, which I suppose
> anyone else can do too) and the guinea pig I picked ("make pilot") seemed
> to build as expected.
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/istio/istio/pull/2831#issuecomment-360959384>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFI6kauyKr0T6VU1RtGgeiFqB0Y--E4ks5tOqcbgaJpZM4RsdJw>
> .
>

mattdelco commented:
Regarding monolithic vs. helper mk files, my personal preference given the current state of istio would be for a single monolithic Makefile since IMO it's not obnoxiously large, still has relatively independent sections, makes it easier to do a single search for uses of a variable/macro/etc, and would make it easier to implement the self-docing-Makefile feature that rshriram requested.

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | a41d7459681f4f2b7f678c0a1f40930eee4efec9 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3566/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 6e88c7d9ab3b03dae5fcbe01326b373cacaf3769 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3579/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

ldemailly commented:
failure is strange : build timeout ?

/test istio-presubmit

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 6e88c7d9ab3b03dae5fcbe01326b373cacaf3769 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3586/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

mattdelco commented:
> failure is strange : build timeout ?

Looks like the presumit gets killed after 60 min.  Of that it was 1.5 min of init, 5 min for "make build", 24 min for "make localTestEnv test" and the remaining 30 min for "make push".  I didn't spot any obvious time hog--offhand it looks like things overall were running slow (e.g., 250MB pushes would take 1-2 min each).  We'll see what happens when I rebase and that causes the tests to trigger again.

ldemailly commented:
we're still only pushing to gcr.io right ? (and not trying docker which would explain the slowness)

mattdelco commented:
I've only seen the prow-based tests use gcr.io/istio-testing (CircleCI seems to only use the build system's local registry).  There's some potential to improve/optimize how [and how often] the tests are invoking make (or other scripts directly [e.g. init.sh]), but for now I'm trying to deter that to a separate effort.

ldemailly commented:
looks like you need to manually rebase because of the conflict, I'll re /lgtm once you're done

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-smoke.sh | 8ae70856d6cd991104a4abf69eec18cc8c0d6bd6 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/2023/) | `/test e2e-smoke`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

rkpagadala commented:
/test e2e-smoke

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-smoke.sh | 8ae70856d6cd991104a4abf69eec18cc8c0d6bd6 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/2031/) | `/test e2e-smoke`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

mattdelco commented:
I'm not clear on whether the "Error: unknown flag: --livenessProbePath" is breakage from a PR that went in a few hours ago, and/or something only I might be seeing as a consequence of changing prow/istio-pilot-e2e.sh to set "-mixer=true" (rather than false), which was done as instructed to get more consistency between what CircleCI and prow are doing.  In my investigation I got concerned that e2e-suite.sh is throwing away parameters like --auth_enable that are passed it by other scripts like e2e-suite-rbac-auth.sh (also --cluster_wide  by e2e-suite-rbac-auth.sh ).

rkpagadala commented:
I think it is because of the change "-mixer=true". I think testing mixer in pilot e2e is legacy from before the mono repo change @kyessenov

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-smoke.sh | 5ce13b2af79c126cd396c29e8cecd530fc70cdba | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/2038/) | `/test e2e-smoke`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

kyessenov commented:
Not sure why I'm mentioned. Yes, there used to be a mock mixer in pilot that was used instead of real mixer. mixer=false disables mixer-related tests altogether, which is something not desirable.

Thanks for the cleanup, by the way. You seem to find your way in Makefiles much better than me. I also vote for a single monolithic Makefile, that calls out to a bunch of bash/python. Bash/python can be split into libraries.

ldemailly commented:
I saw the error in another pr too, andra saw it and cc @jmuk

mattdelco commented:
The failures I saw with -mixer=true seemed to be about a newly added parameter:

Error: unknown flag: --livenessProbePath

With -mixer=false the failures I see mostly remind me of when we were all trying to cope with the crash-prone proxy.

I had this change passing tests last week, and again during the weekend (after rebasing introduced some issues), but today's rebasing seems less straight-forward to address.  I've had certain doubts about some aspects of the tests and punted those over to a different PR to see if things improve at all.

istio-testing commented:
@mattdelco: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-smoke.sh | 5ce13b2af79c126cd396c29e8cecd530fc70cdba | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/2038/) | `/test e2e-smoke`
prow/istio-pilot-e2e.sh | b06e42b2375aad585f61b7615d274889c252d994 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/2014/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

rshriram commented:
Matt if you are confident that these failures are not from your pr, let’s merge this. It would at least help others debug on their respective platforms.

mattdelco commented:
/test istio-pilot-e2e
The test failure swapped from one to another (leftmost of the following) despite me making essentially no code change:
https://k8s-gubernator.appspot.com/pr/istio/istio/2831
Generatelly speaking I've seen lots of failures in the majority of my PRs, and any failure that didn't involve a simple issue (e.g. missing file) has nearly always been due to test flakiness or another change.  Having said that, I'd still like to get this all passing so I'm not stuck firefighting in case this turns out to be the one exception.  This change also affects the way that most people build (either because I deleted their script/Makefile, or because I changed where make places the artifacts) so I should also send an email first to the dev mailing list that explains how to map the old ways to the new.

mattdelco commented:
Looks like all that's missing is my email and a lgtm, assuming there's not other parties I should also check with.  The other reason I had a hold is because I didn't want to get in the way of the 0.5.0 (and though it has a RC I don't want to get in the way if a last-minute fix/rebuild is needed).

jeffmendoza commented:
The email you sent out was pretty epic. I imagine a lot of that info belongs in master/DEV-GUIDE.md

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | f95ec2d09d7c78bd7b93522f97dfc8bdcc4a4f75 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/2056/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

rkpagadala commented:
/lgtm
/approve

rkpagadala commented:
@ldemailly please add /approve

rshriram commented:
/lgtm
/approve

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | e03f3fd7e0ed9352220483d2c45e0a0a63bb72ae | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-presubmit/3752/) | `/test istio-presubmit`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

rkpagadala commented:
/test istio-presubmit

istio-testing commented:
@mattdelco: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | e03f3fd7e0ed9352220483d2c45e0a0a63bb72ae | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/2102/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | e03f3fd7e0ed9352220483d2c45e0a0a63bb72ae | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/2102/) | `/test istio-pilot-e2e`
prow/e2e-smoke.sh | e03f3fd7e0ed9352220483d2c45e0a0a63bb72ae | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/2130/) | `/test e2e-smoke`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

rkpagadala commented:
/test e2e-smoke
/test istio-pilot-e2e

istio-testing commented:
@mattdelco: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-smoke.sh | e03f3fd7e0ed9352220483d2c45e0a0a63bb72ae | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/2130/) | `/test e2e-smoke`
prow/istio-pilot-e2e.sh | e03f3fd7e0ed9352220483d2c45e0a0a63bb72ae | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/2113/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | e03f3fd7e0ed9352220483d2c45e0a0a63bb72ae | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/2113/) | `/test istio-pilot-e2e`
prow/e2e-smoke.sh | e03f3fd7e0ed9352220483d2c45e0a0a63bb72ae | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/2142/) | `/test e2e-smoke`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

rkpagadala commented:
/retest

istio-testing commented:
@mattdelco: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | e03f3fd7e0ed9352220483d2c45e0a0a63bb72ae | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/2113/) | `/test istio-pilot-e2e`
prow/e2e-smoke.sh | e03f3fd7e0ed9352220483d2c45e0a0a63bb72ae | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/2150/) | `/test e2e-smoke`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-smoke.sh | e03f3fd7e0ed9352220483d2c45e0a0a63bb72ae | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/2150/) | `/test e2e-smoke`
prow/istio-pilot-e2e.sh | e03f3fd7e0ed9352220483d2c45e0a0a63bb72ae | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/2124/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

ldemailly commented:
wait for #3016 to retest

andraxylia commented:
/test all

andraxylia commented:
/lgtm

istio-testing commented:
@mattdelco: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-pilot-e2e.sh | e03f3fd7e0ed9352220483d2c45e0a0a63bb72ae | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/2124/) | `/test istio-pilot-e2e`
prow/e2e-smoke.sh | 2702825103689ce696366c262d51a6456191d0a3 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/2166/) | `/test e2e-smoke`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@mattdelco: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/e2e-smoke.sh | 2702825103689ce696366c262d51a6456191d0a3 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/e2e-smoke/2166/) | `/test e2e-smoke`
prow/istio-pilot-e2e.sh | 2702825103689ce696366c262d51a6456191d0a3 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/2831/istio-pilot-e2e/2139/) | `/test istio-pilot-e2e`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->
